[RPC] Use leaky singleton for current RPC agent global

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#41591 [RPC] Use leaky singleton for current RPC agent global**

This allows to delay its construction until it's first used and, therefore, ensure its destruction before all global static objects. In particular, if we didn't do this, and if the user forgot to explicitly call `rpc.shutdown()`, then the static destruction order could first destroy the glog mutexes and only then destroy the global static shared_ptr holding the RPC agent, which would only shut down at that moment, and if it were to encounter an error and attempt to log it would fail and crash the process.

Differential Revision: [D22597100](https://our.internmc.facebook.com/intern/diff/D22597100/)

lw committed:
[RPC] Use leaky singleton for current RPC agent global

This allows to delay its construction until it's first used and, therefore, ensure its destruction before all global static objects. In particular, if we didn't do this, and if the user forgot to explicitly call `rpc.shutdown()`, then the static destruction order could first destroy the glog mutexes and only then destroy the global static shared_ptr holding the RPC agent, which would only shut down at that moment, and if it were to encounter an error and attempt to log it would fail and crash the process.

Differential Revision: [D22597100](https://our.internmc.facebook.com/intern/diff/D22597100/)

[ghstack-poisoned]

jjlilley commented:
> This allows to delay its construction until it's first used and, therefore, ensure its destruction before all global static objects

Is construction time relevant here? 
In both cases, the variable is declared as module-wide static, and the result shouldn't be wildly different. (Maybe it will be in practice, but the results will then be brittle)

The issue here stems from cross-module destructor ordering, which we don't have direct easy control over. Typically, the most practical approaches are to either leak the (static RpcAgent) var, or eliminate the dependency.

lw commented:
> Is construction time relevant here?

Honestly, I'm still not super clear about the workings of static variables. But I found this on [cppreference.com](https://en.cppreference.com/w/cpp/utility/program/exit), which seems to hint that this approach may work:

> If the completion of the constructor or dynamic initialization for thread-local or static object A was sequenced-before thread-local or static object B, the completion of the destruction of B is sequenced-before the start of the destruction of A

lw committed:
Update on "[RPC] Use leaky singleton for current RPC agent global"


This allows to delay its construction until it's first used and, therefore, ensure its destruction before all global static objects. In particular, if we didn't do this, and if the user forgot to explicitly call `rpc.shutdown()`, then the static destruction order could first destroy the glog mutexes and only then destroy the global static shared_ptr holding the RPC agent, which would only shut down at that moment, and if it were to encounter an error and attempt to log it would fail and crash the process.

Differential Revision: [D22597100](https://our.internmc.facebook.com/intern/diff/D22597100/)

[ghstack-poisoned]

pritamdamania87 commented on torch/csrc/distributed/rpc/rpc_agent.cpp:
```diff
@@ -236,14 +238,20 @@ const WorkerInfo& RpcAgent::getWorkerInfo() const {
+  #ifdef C10_USE_GLOG
+    google::base::GetLogger(google::GLOG_ERROR);
+  #endif
+  static std::shared_ptr<RpcAgent> currentRpcAgent = nullptr;
```
This isn't a leaky singleton, you probably want something like `static RpcAgent* currentRpcAgent` so that it leaks. A full example can be found here: https://isocpp.org/wiki/faq/ctors#static-init-order-on-first-use.

Also, in general it's fragile to try to order initialization of static variables and a refactor could easily break this. We should use a leaky singleton as mentioned above.

lw replied:
Yeah, I think here we were in fact going for a different idiom: https://isocpp.org/wiki/faq/ctors#construct-on-first-use-v2. I used the wrong name in the PR summary.

With a true leaky singleton approach the destructor of the agent would _never_ be called, right? I am slightly worried that this may cause problems in some rare cases. For one, the event loop threads would never be joined and would just be destroyed by the OS. I suspect this means that if an error occurs (coming from an external source, like another worker shutting down) during shutdown after the glog mutexes have been destroyed then those threads would still be up and would still attempt to log a message, failing for the same reason. This would only be fully safe if those threads were down by the time the mutexes are destroyed. Also, by never destroying the agent we also never destroy the TensorPipe context, which may mean we may not perform some cleanup operations. At the moment that's not a problem, because all the resources used by TensorPipe are "ephemeral" and will be collected by the kernel, but this may not remain the case forever for future transports.

pritamdamania87 reviewed (approved):
Thanks for fixing this! I think this is probably the best way to resolve this since as you mentioned the leaky singleton could cause other issues since there would be threads alive that might reference destroyed glog mutexes.

lw commented:
It's actually not yet clear whether this fix actually works. @Flamefire have you had a chance to test it? Note that the failing tests have been fixed by adding `rpc.shutdown()` so to test this fix you'll have to comment that out or revert that commit.

Flamefire commented:
I think I already did and it failed: https://github.com/pytorch/pytorch/issues/41561#issuecomment-660259726

However I tried this again: Applied the Patch on the 1.6 tree, so without the rpc.shutdown, rebuild and ran the 2 tests from https://github.com/pytorch/pytorch/issues/41474 and got failures in 2 out of 7 invocations of the test. I captured a backtrace again and it looks the same as https://github.com/pytorch/pytorch/issues/41474#issuecomment-659236045 so I guess it's still this issue.

lw commented:
> I think I already did and it failed: #41561 (comment)

I had updated this PR after that attempt. Thanks for trying again and letting us know this still doesn't work. I'm running out of ideas on how to force a dependency at shutdown so I'm going to close this issue and say that for now we require users explicitly calling `rpc.shutdown()`.
